### ISSUE TITLE
Copyright assignment

### ISSUE THREAD

**Issue by Thrameos** [jpype-project/jpype#545]
@marscher would it be possible to transfer the copyright to some neutral agency such as Apache?  It just feels silly maintaining the Steves original copyright 2004 statement when 90%+ of the code has been rewritten.  (Not to mention the awkwardness of that darn accented e getting crushed by editors).  

I am fine with doing the paperwork and signing it over my portions, but we need to get Steve to sign off and find a organization to accept it.

**Comment by marscher** [comment]
Even if 90% of the code has been rewritten, it is still considered a derivative in terms of clause 2 of the Apache license. What is the benefit of this administrative work, I can now only see that you want to delete the copyright notice in the file headers. Does Apache foundation provide some beneficial framework in the sense of lowering maintenance burdens? We do not need to pay for continuous integration services, hosting etc.

What is the implication of changing the copyright notice anyways? The license still applies and it grants enough freedom in my opinion.

Maybe related https://github.com/jpype-project/jpype/issues/11

**Comment by Thrameos** [comment]
My main issue is in the creation of new files.  I always perfer to assign the copyright to some neutral agency with has the resources to defend rather than retain it myself.  Thus far with most times I have just been copying the projects archival 2004 copyright so long as there is a single line which should apply to most of native/common.  Native/python is currently entirely replaced as of 0.8.  The Jpype python is mixed. 

There is no read guidance in the docs on the project.  Should new authors copy the old header, or should they insert their own copyright statement?

**Comment by marscher** [comment]
Did you ever need to defend your copyright against somebody? I thought it is something based on the U.S. constitution, so it is also only applicable there. Honestly I do not care so much, as we have the git history, which can be used to prove authorship if even questionable. The headers has become irrelevant in my opinion and are pre-git historic. If you think it is worth the effort to transfer this right to an trustworthy agency, please go ahead.

**Comment by Thrameos** [comment]
The other minor wrinkle regarding copyright is that to create the 0.8 speed patch without any documentation from cpython regarding their api i had to read cpython source code and use it as a template.  Both projects have free licenses, but they are different licenses thus giving me hesitation.  

I am sure that cpython authors would not object as the purpose is to promote python in general.  But be assigning the files to Steve I feel like I am dragging a person into something.  It is not clear to me if I should add a statement like "Portions of this code were derived from the CPython source, copyright blah blah... see PSF license".  

Last time i had issues like this I just worked with Stallman and decide to assign the whole project over to Gnu.

**Comment by marscher** [comment]
So you're saying that you've copied over parts of CPython? Have you rewritten it somehow? If so it should be considered your work. Mixing licenses should be avoided in any case.

**Comment by marscher** [comment]
Now it seems to me that this cpython copied/derived code is your main concern about this whole copy-right/right-transferal issue. If the Python foundation would sue this project for some copied lines of code, the GNU foundation would have to deal with it. Ultimately leading to a waste of every bodies resources (lawyers, courts, finally developers needing to re-implement the questionable lines of code).

**Comment by Thrameos** [comment]
Well the only party that I am really worried about is my employer.  I unfortunately work under Department of Energy in the United States in California contracted through a private company (who is replaced every few years).  Virtually everything I write it thus owned by my employer. It was better when my contractor was University of California, but when we were forced to go private more issues have ensued.  My current employer has not caused issues, but at some point they may be replaced by something like Lockheed Martin.   I have been a party to copyright dispute for the last 7 years between DOE, my company, and the funding agency which has prevented a code that I authored from ever seeing the light of day.   The last thing I want to see is a repeat of that whole mess.   

Thus unfortunately I have seen what happens.  I have informal permission to work on this project on my own time from my boss. Filing paperwork with my employer to sign over my portions to a neutral agency just takes the whims of my employer out of the picture, but I can't really file paperwork to sign over to Steve.  Thus always leaving me and the project subject to the whims of whatever the United States Government decides our next private contractor will be.

**Comment by marscher** [comment]
On 12/16/19 3:59 PM, Karl Nelson wrote:
> Well the only party that I am really worried about is my employer. I 
> unfortunately work under Department of Energy in the United States in 
> California contracted through a private company (who is replaced every 
> few years). Virtually everything I write it thus owned by my employer. 
> It was better when my contractor was University of California, but when 
> we were forced to go private more issues have ensued. My current 
> employer has not caused issues, but at some point they may be replaced 
> by something like Lockheed Martin. I have been a party to copyright 
> dispute for the last 7 years between DOE, my company, and the funding 
> agency which has prevented a code that I authored from ever seeing the 
> light of day. The last thing I want to see is a repeat of that whole mess.
> 
> Thus unfortunately I have seen what happens. I have informal permission 
> to work on this project on my own time from my boss. Filing paperwork 
> with my employer to sign over my portions to a neutral agency just takes 
> the whims of my employer out of the picture, but I can't really file 
> paperwork to sign over to Steve. Thus always leaving me and the project 
> subject to the whims of whatever the United States Government decides 
> our next private contractor will be.

What a mess. Thanks for this detailed explanation. I hope we can 
convince Steve (@DevilWolf) in this case to hand over the copy-right to 
a neutral agency.

**Comment by Thrameos** [comment]
Well I am not sure that we have to have Steve hand over the code.  I just need a neutral agency to accept the project so that I can transfer my portions to them thus taking my employer (or any future company that takes over my contract) for having any claim.  Usually that requires prior authors approval but I am not sure it that has to be the case.

With regard to the CPython source.  I had to use  ``Modules/_testcapimodule.c`` as an example for setting up heap classes as well reviewing  ``Object/typeobject.c``, ``Object/exceptions.c`` and ``Object/longobject.c``.  No code was copied directly but I did need to view the source and set up tests to determine behavior and interactions prior so that I could replicate similar functionality within JPype.  The lack of a _noody_ example for the API meant delving into code of another license which always prefer not to do.   (I did discover a few bugs in CPython that I likely should report.)

**Comment by originell** [comment]
Not sure it's the right organization, but maybe someone at the EFF can help out here?  Looks like some legal counseling might be adviseable. Especially because the code is internationally developed.

**Comment by Thrameos** [comment]
I talked it over with my local company lawyer and was directed work through my IPO office.  We have policy called “Incremental contributions to external open source projects” which allows me to freely contribute bug fixes and the like to projects.  Unfortunately there were two sticking points,  first that my contribution can’t exceed more than 25% of the total code in project and the license be on the list of approved  open source licenses of which “Apache” is not included.  As I have exceeded this limit I have to go through the formal approval process and get the appropriate waver.  This usually requires adding the DOE disclaimer and auspices statement as well as attaching the copyright statement of my organization to the project for files that I have contributed.  It will not alter the license nor change the way in which project can be used in any way, but it is a bit of extra red tape.  Once that is complete it should shield the project from any future claims should our management change.  The only part that I am hoping to avoid is the requirement for “review and release” on patches which would very much slow down my ability to contribute if I have to push each patch through a week long review with half a dozen approvers.

Does anyone have any objections to that?
